### PR TITLE
Benchmark scripts improvements

### PR DESCRIPTION
**Pull request by albestro**
As per discussion, this PR for sharing ongoing works on the benchmark scripts side.

The seed of this PR is a set of 4 scripts `gen_{chol/trsm_lln}_{weak/strong}` which are partial and a bit cluttered (sorry, I didn't have the time and I thought it would have been better to start sharing with you at least a starting point).

Feel free to use them as a starting point and change them according to new needs.

@msimberg @teonnik @Sely85

### CODE REVIEW

**Comment by msimberg** [comment]
I've pushed a bunch of miscellaneous changes. They're not particularly well-tested, so apologies in advance if I've missed some corner cases. In no particular order:
- The required command line flags for GPU execution with slate and dplasma are now added automatically (there's a GPU field in the system definitions for this).
- Fixed the scoping issue in the triangular solver miniapp (see #411).
- Added hegst generator and plotting scripts.
- Added command line options to generator and plotting scripts.
- Removed old scripts (e.g. `gen_chol.py`, there's only `gen_chol_weak.py` and `gen_chol_strong.py` now).

The command line options for the generator scripts are:
- `--debug`: prints the job text and does not submit the job
- `--libs [lib1] [lib2] ...`: submit jobs for the given libs

The plotting scripts just have a `--path` flag now that defaults to the current directory.

Definitely to do before we merge this:
- Harmonize benchmark naming (e.g. hegst, gen2std, gen_to_std), I've currently made no attempt to make them consistent.
- Move common argument parsing to a separate file.

**Comment by rasolca** [comment]
@teonnik @albestro @msimberg  Based on the discussion we had I started refactoring the generation scripts.
I concentrated on the strong scaling scripts: please have a first look on the new `StrongScaling` utility and to `gen_strong.py` examples.
Based on the comments I will then work on a similar weak scaling utility.

**Review by albestro — COMMENTED** [review]
Well done! I still haven't had the chance to try it on the field, but it looks very flexible!

Just checking my understanding, and possibly suggesting a way to document it. Please correct me If I'm wrong.

`StrongScaling` acts as a "template" for slurm jobs, where the template is parameterized over `nnodes_arr`.

With its `add` method we can add to each instance of the template (a job), a set of job steps generated by unrolling the parameters (obtained as Cartesian product of them).

With `jobText` we "instantiate" all job instances and get their scripts, while with `submit` we can actually launch each of this job using `sbatch`.

**Comment by rasolca** [comment]
> Well done! I still haven't had the chance to try it on the field, but it looks very flexible!
> 
> Just checking my understanding, and possibly suggesting a way to document it. Please correct me If I'm wrong.
> 
> `StrongScaling` acts as a "template" for slurm jobs, where the template is parameterized over `nnodes_arr`.
> 
> With its `add` method we can add to each instance of the template (a job), a set of job steps generated by unrolling the parameters (obtained as Cartesian product of them).
> 
> With `jobText` we "instantiate" all job instances and get their scripts, while with `submit` we can actually launch each of this job using `sbatch`.

Yes, correct.

In the meantime I also worked on plotting script, adding a couple of features and reducing code duplication.

File: scripts/postprocess.py [hunk 1]
```diff
@@ -269,7 +276,9 @@ def _parse_line_based(fout, bench_name, nodes):
 # │   ├── <bench_name_2>.out
 # |   ...
 #
-def parse_jobs(data_dirs):
+# If distinguish_dir is True the bench name is prepended with the directory name
```
Comment (albestro) [inline scripts/postprocess.py:279]:
```suggestion
# If distinguish_dir is True, bench name is prepended to data directory name (e.g. bench_name@data_dir)
```

**Review by albestro — COMMENTED** [review]
LGTM.

**Comment by msimberg** [comment]
> @teonnik @albestro @msimberg Based on the discussion we had I started refactoring the generation scripts. I concentrated on the strong scaling scripts: please have a first look on the new `StrongScaling` utility and to `gen_strong.py` examples. Based on the comments I will then work on a similar weak scaling utility.

Looks like a huge improvement to me! I'm fine with this direction. Thanks for cleaning these up!

**Comment by teonnik** [comment]
I added a configuration for Cineca's M100 cluster to `systems.py`. The clean-up looks good to me!

**Review by Sely85 — APPROVED** [review]

File: scripts/systems.py [hunk 2]
```diff
@@ -72,8 +89,72 @@
...
+    "Batch preamble": """
+#!/bin/bash -l
+#SBATCH --job-name={run_name}_{nodes}
+#SBATCH --time={time_min}
+#SBATCH --nodes={nodes}
+#SBATCH --ntasks-per-node={rpn}
+#SBATCH --cpus-per-task={threads_per_rank}
+#SBATCH --partition=m100_usr_prod
+#SBATCH --account=cin_staff
+#SBATCH --qos=normal
```
[the remark this answers is not in the document]
Comment (rasolca) [reply scripts/systems.py:119]: Is needed to have multiple qos or `m100_qos_bprod` for any number of nodes is good?

[the remark this answers is not in the document]
Comment (Sely85) [reply scripts/systems.py:119]: It is needed to have multiple qos: `normal` should be used for maximum 16 nodes, while `m100_qos_bprod` is needed when asking for a number of nodes in range [17-256].

**Review by Sely85 — APPROVED** [review]
LGTM. Marconi100 looks "correct" now (following US hints, at least).

**Pull request merged**